File: ios/TerminalDeck/Screens/MachineToolsSection.swift
/**
 * The six doors into a machine that the desktop has always had and the phone
 * did not: **Files, Source control, Artifacts, Store, AI readiness, MCP
 * servers**.
 *
 * > *"what about files, artifacts, source control, store, ai readiness, mcp
 * > servers in ios app too for server"*
 *
 * > *"all what i asked for so many times, i need all no exceptions."*
 *
 * All six, in one section, in the order he said them — because the order a
 * person asked for things in is the order they look for them in, and reordering
 * them by capability group would put Source control between Store and MCP for
 * no reason a reader could see.
 *
 * ## A section rather than a screen
 *
 * It is drawn on `MachineDetailView` today, which is the page about one machine.
 * It is a section and not a screen so the same six rows can also hang off a
 * server's page, or a session's folder, without a second copy of the list
 * drifting out of step with this one.
 *
 * ## Three capabilities gate six rows, and an absent row is the answer
 *
 * `files`, `git` and `panels` are all **owner-device only** on the host — a
 * guest holding them could read a private key out of a folder it was lent, or a
 * credential out of an MCP config. A phone that is not the owner's therefore
 * gets a shorter list, and that is the app working.
 *
 * A row is **not drawn** rather than drawn and refused, which is the rule every
 * gated control in this app follows: a button whose only outcome is a refusal
 * is not a button. The reason is not guessed at here either — `MachineDetailView`
 * prints the machine's own capability list a few sections up, so somebody
 * surprised by a missing row has the machine's answer rather than a hunch about
 * their own device kind.
 *
 * ## Why the two destinations come in as closures
 *
 * Files and Source control are their own screens, owned by their own lanes;
 * the four panels are `PanelView` and are built here. Taking the first two as
 * `@ViewBuilder` closures keeps this file from depending on the exact name and
 * initialiser of two screens being written beside it — the caller, which knows
 * both, says what to push. It is also the honest seam: this section owns *the
 * list and its gates*, not what is behind each row.
 */

import SwiftUI

struct MachineToolsSection<FilesDestination: View, SourceDestination: View>: View {
    let model: DeckModel
    /**
     * The folder the six doors open onto.
     *
     * `nil` is a real answer and not a missing one: a machine that has granted
     * this phone no folder still has a Store and a readiness check, and the host
     * answers `panel.read` with no path from its own default. Files and Source
     * control are the two that genuinely need one, and the closures that build
     * them decide what to do about it — they were handed the same value.
     */
    var path: String?

    @ViewBuilder let filesDestination: () -> FilesDestination
    @ViewBuilder let sourceDestination: () -> SourceDestination

    var body: some View {
        // Nothing at all when no door is open — not a caption over an empty
        // card, which reads as a section that failed to load.
        if model.canReadFiles || model.canReadGit || model.canReadPanels {
            caption
            card {
                if model.canReadFiles {
                    link("Files", icon: "folder", id: "files") { filesDestination() }
                }
                if model.canReadGit {
                    if model.canReadFiles { line }
                    link("Source control", icon: "arrow.triangle.branch", id: "git") { sourceDestination() }
                }
                if model.canReadPanels {
                    if model.canReadFiles || model.canReadGit { line }
                    panelLink("Artifacts", icon: "shippingbox", panel: .artifacts)
                    line
                    panelLink("Store", icon: "square.grid.2x2", panel: .store)
                    line
                    panelLink("AI readiness", icon: "checkmark.seal", panel: .readiness)
                    line
                    panelLink("MCP servers", icon: "point.3.connected.trianglepath.dotted", panel: .mcp)
                }
            }
        }
    }

    // MARK: - Rows

    /// One of the four panels, all of which are the same screen with a different
    /// id and a different word on the title bar. See `PanelView`.
    /// `PanelKind` rather than a string: a panel this build cannot draw becomes a
    /// compile error here rather than a refused frame, and a refused frame closes
    /// the socket — which reads to a person as the network dropping.
    private func panelLink(_ title: String, icon: String, panel: PanelKind) -> some View {
        link(title, icon: icon, id: panel.rawValue) {
            PanelView(panel: panel, title: title, model: model, path: path)
        }
    }

    /**
     * A row, at the settings screen's own two numbers.
     *
     * A `NavigationLink` with a destination rather than a routed value, which is
     * what `LocalhostListView` and `WatchView` already do for pushed screens
     * that no other surface deep-links to. Six new cases on
     * `DeckModel.SettingsRoute` would each need a `navigationDestination` arm
     * inside `DeckTabs`, and none of the six is ever pushed from anywhere but
     * this list.
     */
    private func link<Destination: View>(
        _ title: String,
        icon: String,
        id: String,
        @ViewBuilder destination: @escaping () -> Destination
    ) -> some View {
        NavigationLink {
            destination()
        } label: {
            HStack(spacing: 12) {
                /*
                 * Monoline: 19 points at light weight in a 24-point column.
                 * SF Symbols' default is regular at 15, which is what every iOS
                 * app that has not thought about it looks like — the exact
                 * complaint. `SettingsRowBody` carries the full argument.
                 */
                Image(systemName: icon)
                    .font(.system(size: 19, weight: .light))
                    .foregroundStyle(Theme.secondary)
                    .frame(width: 24)
                Text(title)
                    .font(.system(size: 16))
                    .foregroundStyle(Theme.primary)
                Spacer(minLength: 8)
                Image(systemName: "chevron.right")
                    .font(.system(size: 13, weight: .semibold))
                    .foregroundStyle(Theme.faint)
            }
            .padding(.horizontal, 16)
            .padding(.vertical, 13)
            .frame(maxWidth: .infinity, alignment: .leading)
            .contentShape(Rectangle())
        }
        .buttonStyle(.plain)
        .accessibilityIdentifier("machine.tools.\(id)")
    }

    // MARK: - Its own chrome

    /*
     * Drawn here rather than borrowed: `SectionCaption`, `SettingsGroup` and
     * `SettingsDivider` are **private to `DeckTabs.swift`**, so a screen that
     * reached for them would have to live in that file — the argument
     * `AppLockScreen` and `MachineDetailView` both make about the same three
     * names. Same metrics: an 11pt kerned caption, a `Theme.surface` card at
     * radius 20, and a hairline inset to the label.
     *
     * 52 here rather than `MachineDetailView`'s 16, and it is arithmetic: these
     * rows have an icon column, so the words start at 16 + 24 + 12. A line at 16
     * would cut through the gutter the icons stand in.
     */
    private var caption: some View {
        HStack(spacing: 4) {
            Text("LOOK INSIDE")
                .font(.system(size: 11, weight: .semibold))
                .kerning(0.6)
                .foregroundStyle(Theme.faint)
            // The one sentence this section is allowed, and it is behind the ⓘ:
            // *"I don't want any kind of long descriptions anywhere."* It earns
            // its place because read-only is the non-obvious half — six rows
            // that look like the desktop's panels, none of which will change
            // anything on the machine.
            InfoDot(
                about: "Look inside",
                text: "Everything here is read-only. These six read the machine — its folders, "
                    + "what git says, and its four panels — and none of them change anything on it."
            )
            Spacer(minLength: 0)
        }
        .padding(.leading, 4)
        .padding(.top, 24)
        .padding(.bottom, 8)
        .frame(maxWidth: .infinity, alignment: .leading)
    }

    private func card<Content: View>(@ViewBuilder _ content: () -> Content) -> some View {
        VStack(spacing: 0) { content() }
            .background(Theme.surface, in: RoundedRectangle(cornerRadius: 20, style: .continuous))
    }

    private var line: some View {
        Rectangle()
            .fill(Theme.hairline)
            .frame(height: 0.5)
            .padding(.leading, 52)
    }
}

File: android/app/src/main/java/dev/terminaldeck/android/ui/MachineToolsSection.kt
package dev.terminaldeck.android.ui

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.KeyboardArrowRight
import androidx.compose.material.icons.filled.AccountTree
import androidx.compose.material.icons.filled.Folder
import androidx.compose.material.icons.filled.GridView
import androidx.compose.material.icons.filled.Hub
import androidx.compose.material.icons.filled.Inventory2
import androidx.compose.material.icons.filled.Verified
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.mutableStateListOf
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.unit.dp
import androidx.compose.ui.window.Popup
import androidx.compose.ui.window.PopupProperties
import dev.terminaldeck.android.DeckUiState
import dev.terminaldeck.android.FilesGitController
import dev.terminaldeck.android.PanelsController
import dev.terminaldeck.android.protocol.GitFileChange
import dev.terminaldeck.android.protocol.PanelKind
import dev.terminaldeck.android.tunnel.TunnelView
import dev.terminaldeck.android.ui.kit.DeckDivider
import dev.terminaldeck.android.ui.kit.DeckGroup
import dev.terminaldeck.android.ui.kit.DeckQuietButton
import dev.terminaldeck.android.ui.kit.InfoDot
import dev.terminaldeck.android.ui.theme.DeckTheme
import dev.terminaldeck.android.ui.theme.DeckType
import dev.terminaldeck.android.ui.theme.Space

/**
 * The "look inside" doors into a machine — a port of `ios/TerminalDeck/Screens/MachineToolsSection.swift`.
 *
 * The six rows the desktop has always had and the phone did not — **Files, Source control, Artifacts,
 * Store, AI readiness, MCP servers** — in the order he asked for them. `files`, `git` and `panels` are
 * all owner-device only on the host, so a phone that is not the owner's gets a shorter list, and a row
 * is **not drawn** rather than drawn and refused: a button whose only outcome is a refusal is not a
 * button. Everything behind these six is read-only.
 *
 * ## A section that is also its own nav hub
 *
 * On iOS these are `NavigationLink`s onto the app's own stack. Here the machine screen this drops into
 * must not be edited, so the section owns navigation itself: tapping a row pushes onto an internal
 * stack, and the pushed screen is drawn full-screen in a [Popup] above everything — Files opening the
 * file viewer, Source control opening a diff, the artifacts panel opening [ArtifactView], and the
 * artifact opening back into Files. The system back gesture pops that stack while it is non-empty and
 * leaves the machine screen only once it is empty.
 *
 * @param state the selected machine's ui state — the three capability gates, [DeckUiState.machineNoun],
 *   [DeckUiState.startableFolders], the tunnel and whether ports may be served here.
 * @param files the selected machine's files/git controller (`DeckViewModel.filesGit()`), or null.
 * @param panels the selected machine's panels controller (`DeckViewModel.panels()`), or null.
 */
@Composable
fun MachineToolsSection(
    state: DeckUiState,
    files: FilesGitController?,
    panels: PanelsController?,
    onServePort: (Int) -> Unit,
    onCloseServedPort: () -> Unit,
    modifier: Modifier = Modifier,
) {
    if (!state.canReadFiles && !state.canReadGit && !state.canReadPanels) return

    val colors = DeckTheme.colors
    val stack = remember { mutableStateListOf<ToolRoute>() }
    val start = state.startableFolders.firstOrNull().orEmpty()

    Column(modifier = modifier.fillMaxWidth()) {
        Row(
            verticalAlignment = Alignment.CenterVertically,
            modifier = Modifier.fillMaxWidth().padding(start = Space.x1, top = Space.x6, bottom = Space.x2),
        ) {
            Text("LOOK INSIDE", style = DeckType.overline, color = colors.faint)
            InfoDot(
                about = "Look inside",
                text = "Everything here is read-only. These six read the machine — its folders, what git says, " +
                    "and its four panels — and none of them change anything on it.",
            )
        }
        DeckGroup {
            var drawnAny = false
            if (state.canReadFiles) {
                ToolRow("Files", Icons.Filled.Folder) { stack.add(ToolRoute.Files(start)) }
                drawnAny = true
            }
            if (state.canReadGit) {
                if (drawnAny) DeckDivider(startIndent = 52.dp)
                ToolRow("Source control", Icons.Filled.AccountTree) { stack.add(ToolRoute.Source) }
                drawnAny = true
            }
            if (state.canReadPanels) {
                if (drawnAny) DeckDivider(startIndent = 52.dp)
                ToolRow("Artifacts", Icons.Filled.Inventory2) { stack.add(ToolRoute.Panel(PanelKind.Artifacts, "Artifacts")) }
                DeckDivider(startIndent = 52.dp)
                ToolRow("Store", Icons.Filled.GridView) { stack.add(ToolRoute.Panel(PanelKind.Store, "Store")) }
                DeckDivider(startIndent = 52.dp)
                ToolRow("AI readiness", Icons.Filled.Verified) { stack.add(ToolRoute.Panel(PanelKind.Readiness, "AI readiness")) }
                DeckDivider(startIndent = 52.dp)
                ToolRow("MCP servers", Icons.Filled.Hub) { stack.add(ToolRoute.Panel(PanelKind.Mcp, "MCP servers")) }
            }
        }
    }

    if (stack.isNotEmpty()) {
        Popup(properties = PopupProperties(focusable = true)) {
            Box(modifier = Modifier.fillMaxSize().background(colors.background)) {
                val pop: () -> Unit = { if (stack.isNotEmpty()) stack.removeAt(stack.lastIndex) }
                when (val route = stack.last()) {
                    is ToolRoute.Files ->
                        if (files != null) FilesScreen(files, route.start, onOpenFile = { p, s -> stack.add(ToolRoute.FileText(p, s)) }, onBack = pop)
                        else Disconnected(pop)

                    is ToolRoute.FileText ->
                        if (files != null) FileTextScreen(files, route.path, route.size, onBack = pop) else Disconnected(pop)

                    ToolRoute.Source ->
                        if (files != null) SourceControlScreen(files, start, onOpenDiff = { change, staged -> stack.add(ToolRoute.Diff(start, change, staged)) }, onBack = pop)
                        else Disconnected(pop)

                    is ToolRoute.Diff ->
                        if (files != null) DiffScreen(files, route.repoPath, route.change, route.staged, onBack = pop) else Disconnected(pop)

                    is ToolRoute.Panel ->
                        if (panels != null) PanelScreen(
                            controller = panels,
                            panel = route.kind,
                            title = route.title,
                            path = state.startableFolders.firstOrNull(),
                            canReadPanels = state.canReadPanels,
                            onOpenArtifact = { ref, title, project -> stack.add(ToolRoute.Artifact(ref, title, project)) },
                            onBack = pop,
                        ) else Disconnected(pop)

                    is ToolRoute.Artifact ->
                        if (files != null && panels != null) ArtifactView(
                            ref = route.ref,
                            title = route.title,
                            project = route.project,
                            files = files,
                            panels = panels,
                            machineNoun = state.machineNoun,
                            canReadFiles = state.canReadFiles,
                            canReadPanels = state.canReadPanels,
                            canBrowseLocalhost = state.localhostOffered,
                            tunnel = state.tunnel,
                            onServePort = onServePort,
                            onCloseServedPort = onCloseServedPort,
                            onOpenFolder = { folder -> stack.add(ToolRoute.Files(folder)) },
                            onBack = pop,
                        ) else Disconnected(pop)
                }
            }
        }
    }
}

/** Where the six doors lead. Internal to the section, which is the only thing that pushes them. */
private sealed interface ToolRoute {
    data class Files(val start: String) : ToolRoute
    data class FileText(val path: String, val size: Long?) : ToolRoute
    object Source : ToolRoute
    data class Diff(val repoPath: String, val change: GitFileChange, val staged: Boolean) : ToolRoute
    data class Panel(val kind: PanelKind, val title: String) : ToolRoute
    data class Artifact(val ref: ArtifactRef, val title: String, val project: String?) : ToolRoute
}

@Composable
private fun ToolRow(title: String, icon: ImageVector, onClick: () -> Unit) {
    val colors = DeckTheme.colors
    Row(
        verticalAlignment = Alignment.CenterVertically,
        modifier = Modifier.fillMaxWidth().clickable(onClick = onClick).padding(horizontal = Space.x4, vertical = Space.x3),
    ) {
        Icon(icon, contentDescription = null, tint = colors.secondary, modifier = Modifier.size(24.dp))
        Spacer(Modifier.width(Space.x3))
        Text(title, style = DeckType.body, color = colors.primary, modifier = Modifier.weight(1f))
        Icon(Icons.AutoMirrored.Filled.KeyboardArrowRight, contentDescription = null, tint = colors.faint, modifier = Modifier.size(18.dp))
    }
}

/** A pushed screen whose machine went away underneath it — a re-pair or a disconnect mid-navigation. */
@Composable
private fun Disconnected(onBack: () -> Unit) {
    val colors = DeckTheme.colors
    Column(
        horizontalAlignment = Alignment.CenterHorizontally,
        modifier = Modifier.fillMaxSize().padding(Space.x8),
    ) {
        Spacer(Modifier.size(Space.x16))
        Text("Not connected to this machine any more.", style = DeckType.body, color = colors.secondary)
        Spacer(Modifier.size(Space.x5))
        DeckQuietButton(label = "Back", onClick = onBack, modifier = Modifier.width(200.dp))
    }
}
